### DISCUSSION TITLE
What kind of nodes are needed?

### DISCUSSION THREAD

**Discussion by zeffii** [Marcus-Richmond/sverchok-gis-nodes#1] [Ideas]
Hi, first, thank you for the invitation. Integration into Sverchok will be interesting. 

i imagine the work will happen in 3 areas
1. making nodes, 
    - do they need categories or will the set be manageable from one or two main menus
      _- I think one or two main menus is good for now. Eventually, I could see categories being needed, but I'd like to focus on the basic core functions (see below), which probably won't constitute more than a few nodes_
2. implementing the addon integration with sverchok
    - this should the least of the overal effort to get a working integration
      _- This is where I could use the most help. To start, how to get the repo working so that Sverchok recognized the nodes._ 
      _- Also, you've probably noticed when reviewing the scripts that geopandas and some other 3rd party libraries are needed. So, eventually it would be nice if this addon had options to install these libraries, similar to how Sverchok can install scipy, numpy, and other libraries._
3. improving nodes, customizing intergration, adding menus+operators if needed._

I think point 3. is where most time will be spent.

**Comment by zeffii** [comment]
judging by the work sofar i'd imagine you want at least some basic nodes to start

- GIS loader node
   - with filepath selection on the UI + socket
   - with output socket for GIS data
- GIS decomposer
   - accepts the output from the GIS loader node
   - allows user to maybe see the structure
   - allows user to parametrize the way the data is extracted from the structure ( maybe via a script box, like SV ExecNodeMod or SV Formula node )
- GIS coordinate transformer
    - with curvature correction whatever :)

**Reply by zeffii** [reply]
feel free to edit these descriptions :)

**Comment by Marcus-Richmond** [comment]
First, if you haven't already, please see my responses in _italics_ above.

### core nodes

I envision this addon having two core nodes/functions:

1. Importing geometry from a GIS layer
    - This is really three functions: import point layer, import line layer, import polygon layer 
3. Importing attribute data from a GIS layer

I think these two functions should be encompassed by multiple GIS loader nodes, something like:

- Point Loader
- Line Loader
- Polygon Loader
- Attribute Loader

These nodes can start with the code [here](https://github.com/Marcus-Richmond/sverchok-gis-scripts). I think splitting up the functions into separate nodes is best for now, because the existing code relies upon the geopandas library, and I'm not sure if geopandas should be utilized over another library. I've noticed that geopandas isn't very fast with the data I've tested these scripts with. This may be due to the data being large datasets, my computer not being fast, or something else, so maybe geopandas is the way to go. But, there are other libraries (like gdal/ogr) which can read GIS layers and potentially manipulate the data into a format that can be used in Sverchok, and these other libraries may have some perfomance or other benefits. Frankly, I'm most familiar with geopandas so that's why I've started with it. So, I want to avoid writing a complex GIS loader node relying on geopandas, just to find out some other library should be used, and have to rewrite the complex loader node. 

For instance, the point/polygon/attribute import scripts [here](https://github.com/Marcus-Richmond/sverchok-gis-scripts) use geopandas to turn the GIS layer into a json-like dictionary, then extract the geometry or attribute data from that dictionary and append it to a list. Maybe another library like ogr can read the GIS geometry or attribute information as an array faster. 

### looking forward

Having one GIS Loader node to handle geometry and attributes would be nice. This would probably make building the node tree easier too. Also, having a GIS Decomposer to be able to manipulate the data in various ways with various 3rd party libraries should be created eventually . However, I see this as a very advanced node (at least a node that I suspect only advanced users would use), and I'd like to start with the easier to understand nodes first.

**Reply by zeffii** [reply]
> So, I want to avoid writing a complex GIS loader node relying on geopandas

I wouldn't worry too much about that, python only has one sensible database I/O type and that's the Dict, and the reader code (whatever library used) will result (i'm guessing) in a dict. So if there are multiple libraries to choose from, i think they could even coexist or limit to the most versatile. Get something working first, then consider steps to offer alternatives or to optimize.

In Sverchok as soon as a node has completed processing the data, the data is sent to an internal "cache". Nodes downstream from the node will read the cache associated with the node and its socket. 

If a node isn't directly updated via a slider or inputsocket, then it doesn't do anymore processing. In most cases nodes only have to read a file once (this is usually a slow step - for large files) and they remain "dormant" after that. At the moment that's cached in RAM, and that should be fine for most scenarios .. except.. ofcourse ....  big datasets. If this becomes a problem we can use the HD to cache, this will have its own penalties. But usually as you extract the important data, the ways to optimize the storage sizes increase. You won't always be working on the entire dataset.

**Comment by zeffii** [comment]
cool!   ok, i'll sometimes let you know I read stuff and am thinking about it by using the symbol below.

**Reply by zeffii** [reply]
👀

**Reply by zeffii** [reply]
just to let you know your end of the conversation isn't happening in a vacuum

**Reply by Marcus-Richmond** [reply]
I will try to do the same 👀

**Comment by zeffii** [comment]
ok. the attribute loader may already be in Sverchok as the Dictionary node, the GIS file you used for the SNLite question essentially reads the GIS file and turns it into a `dict` like Object which can be queried by syntax similar to 
```python
gis_data = read_gis_from_path(path)   # a wrapper which gets the GEO interface if the path lookup was success
if gis_data:
    out = gis_data[your_key][some_other_key]
else:
    out = {}
```
in this example the GIS reader node could (electively) be passed a few keys to narrow down the data it outputs

**Reply by Marcus-Richmond** [reply]
So the geo interface function is called on a geodataframe (same thing as a pandas dataframe essentially). Could we have the loader node(s) read the gis layer into a geodataframe, and then output that geodataframe through a specialized node in Sverchok? Do we even need a specialized node? Can a (geo)dataframe go through any of the node sockets already in Sverchok?

Very interesting...

**Reply by zeffii** [reply]
I would not make a separate custom Socket class for this, it's a Dictionary for all intents and purposes.

**Reply by zeffii** [reply]
so `SvDictionarySocket` is appropriate.

**Reply by zeffii** [reply]
unless there is some subtlety in the dataframe, offering more than just keyword lookups.

**Comment by zeffii** [comment]
the biggest pain for me will be to get these libraries installed :D  on windows.

**Reply by Marcus-Richmond** [reply]
I understand that too well haha. Maybe eventually I'll add a comprehensive guide to installing geopandas into blender. For now, I wrote down some rough notes on what I ran into during installation here: [https://github.com/Marcus-Richmond/sverchok-gis-scripts/blob/main/README.md](https://github.com/Marcus-Richmond/sverchok-gis-scripts/blob/main/README.md)

**Comment by zeffii** [comment]
i've resolved a local filesystem issue, and can now load the addon.   

let the games begin!

**Reply by Marcus-Richmond** [reply]
Awesome! Glad you can join in on the fun now haha

**Comment by Marcus-Richmond** [comment]
Hi @zeffii, sorry I haven't contributed anything in so long, but wanted to let you know I'm still interested in working on this project. I've been a bit bogged down with work the past few weeks, so was tough to find the motivation/time to keep up with this. However, I am going to try to get back into this soon, hopefully in the next week or two.

I was trying to get back into it this week, but found myself needing to relearn a bit of Sverchok first haha.... As always, I really appreciate the help you provided on this, I'm still looking forward to the sort of things we can make with Sverchok GIS nodes!

**Reply by zeffii** [reply]
at this point the back-end is mostly there, tidied up and ready for you to fill in the nodes you need.
